src/main/java/com/algorithmia/algorithmHandler/AlgorithmHandler.java

```diff
@@ -0,0 +1,107 @@
+
+import java.lang.reflect.Method;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+
+public class AlgorithmHandler<INPUT, OUTPUT, STATE> {
+
+    private ReflectionHelper.DebuggableBifunction<INPUT, STATE, OUTPUT> applyWState;
+    private ReflectionHelper.DebuggableFunction<INPUT, OUTPUT> apply;
+    private Supplier<STATE> loadFunc;
+
+    private Class algorithmClass;
+    private STATE state;
+
+
+    public AlgorithmHandler(Class algorithmClass, ReflectionHelper.DebuggableBifunction<INPUT, STATE, OUTPUT> applyWState, Supplier<STATE> loadFunc) {
+        this.applyWState = applyWState;
+        this.loadFunc = loadFunc;
+        this.algorithmClass = algorithmClass;
+    }
+
+    public AlgorithmHandler(Class algorithmClass, ReflectionHelper.DebuggableBifunction<INPUT, STATE, OUTPUT> applyWState) {
+        this.applyWState = applyWState;
+        this.algorithmClass = algorithmClass;
+    }
+
+    public AlgorithmHandler(Class algorithmClass, ReflectionHelper.DebuggableFunction<INPUT, OUTPUT> apply) {
+        this.apply = apply;
+        this.algorithmClass = algorithmClass;
+    }
+
+    private void Load() {
+        if (this.loadFunc != null) {
+            state = this.loadFunc.get();
+            System.out.println("PIPE_INIT_COMPLETE");
+            System.out.flush();
+        }
+    }
+
+    private void ExecuteWithoutState(RequestHandler<INPUT> in, ResponseHandler out, Function<INPUT, OUTPUT> func) {
+        INPUT req = in.GetNextRequest();
+        while (req != null) {
+            OUTPUT output = func.apply(req);
+            out.writeToPipe(output);
+            req = in.GetNextRequest();
+        }
+    }
+
+    private void ExecuteWithState(RequestHandler<INPUT> in, ResponseHandler out, BiFunction<INPUT, STATE, OUTPUT> func) {
+        INPUT req = in.GetNextRequest();
+        while (req != null) {
+            OUTPUT output = func.apply(req, state);
+            out.writeToPipe(output);
+            req = in.GetNextRequest();
+        }
+    }
+
+    private void Execute(RequestHandler<INPUT> in, ResponseHandler out) {
+        if (this.applyWState != null && this.loadFunc != null) {
+            Load();
+            ExecuteWithState(in, out, this.applyWState);
+        } else if (this.apply != null) {
+            ExecuteWithoutState(in, out, this.apply);
+        } else {
+            throw new RuntimeException("If using an load function with state, a load function must be defined as well.");
+        }
+    }
+
+    private Class<INPUT> GetInputClass() {
+        String methodName;
+        if (this.applyWState != null) {
+            methodName = ReflectionHelper.getMethodName(this.applyWState);
+        } else if (this.apply != null) {
+            methodName = ReflectionHelper.getMethodName(this.apply);
+        } else {
+            throw new RuntimeException("Either Apply(T t) or Apply(T t S s) must be provided to the constructor.");
```

Comment:
again this might not be necessary if we werne't using reflection.  also `T` and `S` aren't the generic types in this class anywhere, not sure if that was intentional or not.  if we did want to do the reflection route this is something that should be validated at the constructor instead

Reply:
This really should never get thrown, as any constructor should implement either apply or applyWithState - this is just a sanity checking error (I didn't make the class final, maybe I should?)

Reply:
the sanity checking wouldn't be needed if it was an interface that you had to implement :shrug:.   it seems like a more common approach e.g. the Handler across all of the languages here - https://docs.aws.amazon.com/lambda/latest/dg/java-programming-model-handler-types.html